### DISCUSSION TITLE
Supporting several versions of GHC, Cabal

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [Homebrew/discussions#2116] [Everyday usage]
It's useful for development to be able to support several versions of Haskell tools, mainly GHC and Cabal.

Can older versions of those (going back to Cabal 3.0 and GHC 8.6) be "brought back"?

**Comment by carlocab** [comment]
We only package maintained versions of software in Homebrew/core (but make exceptions when certain versions are needed by existing formulae), so [Homebrew policy](https://docs.brew.sh/Acceptable-Formulae) precludes bringing back older versions.

If you need older versions of anything Homebrew packages, you can create formulae for them and host these in your own tap. Here are some docs to help you get started:

https://docs.brew.sh/Taps
https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap
https://brew.sh/2020/11/18/homebrew-tap-with-bottles-uploaded-to-github-releases/

You can probably use these taps for inspiration:
https://github.com/lithammer/homebrew-deadsnakes
https://github.com/shivammathur/homebrew-php

If these older versions existed at some point in Homebrew/core, you may be able to use `brew extract` to recover the version of the formula you need. See `brew help extract` for details.

Note that you can currently do `brew install ghc@8.6` for GHC 8.6.

**Comment by ivanperez-keera** [comment]
Well, the problem here is that cabal-install 3.6, which has already been included in brew, is still not stable (it was released just a month ago, see https://hackage.haskell.org/package/cabal-install/changelog). I've reported 2 bugs today, and right now I can't install my packages.

If we just switch to the latest tool as soon as it's out, we are switching a slightly-older-but-till-recently-maintained tool for a completely-new-and-unstable one. Until the bugs are fixed, and that can take a while (cabal has 1.3K bugs open on github), it's going to be a broken setup.

**Reply by SMillerDev** [reply]
> Well, the problem here is that cabal-install 3.6, which has already been included in brew, is still not stable (it was released just a month ago, see https://hackage.haskell.org/package/cabal-install/changelog). 

That's not what upstream thinks, they released it after all.

> I've reported 2 bugs today, and right now I can't install my packages.

That's very unfortunate and we can patch the formula as soon as patches are available.

> If we just switch to the latest tool as soon as it's out, we are switching a slightly-older-but-till-recently-maintained tool for a completely-new-and-unstable one. 

As far as I know that's how brew has worked since it's inception, although it might have gotten better at it over the years.

> Until the bugs are fixed, and that can take a while (cabal has 1.3K bugs open on github), it's going to be a broken setup.

That's really unfortunate, `brew extract` might be able to help you maintain an older version for yourself.

---
I would like to ask this: if every package manager delayed major releases, how is upstream ever going to find bugs?
